src/sensr_client.cpp

```diff
@@ -1,119 +1,134 @@
-  {
+  Client::~Client() {
     for (const auto& listener : listeners_) {
       UnsubscribeMessageListener(listener);
     }
-    StopReconnecting();
+
+    io_context_.stop();
+    if (reconnection_thread_.joinable()) {
+      reconnection_thread_.join();
+    }
   }
 
   void Client::Reconnect() {
     if (!is_reconnecting_) {
-      StopReconnecting();
-      reconnection_thread_ = std::thread([this]() {
-        is_reconnecting_ = true;
-        auto temp = listeners_;
-        size_t reconnection_attempts_ = 0;
-        while (reconnection_attempts_++ < Client::kMaxReconnectTrialCount) {
-          INFO_LOG("Reconnecting...");
-          for (const auto& listener : listeners_) {
-            UnsubscribeMessageListener(listener);
-          }
-          for (const auto& listener : temp) {
-            SubscribeMessageListener(listener);
-          }
-          bool is_output_connected = true;
-          bool is_point_connected = true;
-          if (IsResultListening()) {
-            if (!output_endpoint_->IsConnected()) {
-              is_output_connected = false;
-            }
-          }
-          if (IsPointListening()) {
-            if (!point_endpoint_->IsConnected()) {
-              is_point_connected = false;
-            }
-          }
-
-          if (is_output_connected && is_point_connected) {
-            INFO_LOG("Reconnected!");
-            break;
-          } else if (!is_reconnecting_) {
-            INFO_LOG("Stop reconnection!");
-            break;
-          }
-          std::this_thread::sleep_for(std::chrono::seconds(1));
-        }
-        is_reconnecting_ = false;
+      if (reconnection_thread_.joinable()) {
+        reconnection_thread_.join();
+      }
+
+      is_reconnecting_ = true;
+      asio::post(io_context_, [this]() { reconnection_async(); });
+      reconnection_thread_ = std::thread([this] {
+        io_context_.restart();
+        io_context_.run();
       });
     }
   }
 
-  bool Client::SubscribeMessageListener(const std::shared_ptr<MessageListener>& listener) { 
+  void Client::reconnection_async() {
+    INFO_LOG("Reconnecting...");
+    auto temp = listeners_;
+    for (const auto& listener : listeners_) {
+      UnsubscribeMessageListener(listener);
+    }
+    for (const auto& listener : temp) {
+      SubscribeMessageListener(listener);
+    }
+    bool is_output_connected = true;
+    bool is_point_connected = true;
+    if (IsResultListening()) {
+      if (!output_endpoint_->IsConnected()) {
+        is_output_connected = false;
+      }
+    }
+    if (IsPointListening()) {
+      if (!point_endpoint_->IsConnected()) {
+        is_point_connected = false;
+      }
+    }
+
+    if (is_output_connected && is_point_connected) {
+      INFO_LOG("Reconnected!");
+      is_reconnecting_ = false;
+    } else {
+      asio::steady_timer sleep_timer(io_context_, asio::chrono::seconds(1));
+      sleep_timer.wait();
+      asio::post(io_context_, [this]() { reconnection_async(); });
```

Comment:
`sleep` is optional and posting the job to io_context_ is not necessary because it's already running asynchronously :wink: 
```suggestion
      reconnection_async();
```

Reply:
Thanks, I will change it. :+1: But I'd like to keep sleep timer since it tries reconnection too much and somehow sdk crashes.

Reply:
Alright, I'll leave it up to you.